pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java

```diff
@@ -390,6 +423,78 @@ private List<Schema> getSchemas() {
     return schemas;
   }
 
+  private void addLogicalTables(List<String> paths) {
+    // Subscribe data changes before reading the data to avoid missing changes
+    for (String path : paths) {
+      _propertyStore.subscribeDataChanges(path, _zkLogicalTableChangeListener);
+    }
+    List<ZNRecord> znRecords = _propertyStore.get(paths, null, AccessOption.PERSISTENT);
+    for (ZNRecord znRecord : znRecords) {
+      if (znRecord != null) {
+        try {
+          putLogicalTable(znRecord);
+        } catch (Exception e) {
+          LOGGER.error("Caught exception while adding logical table for ZNRecord: {}", znRecord.getId(), e);
+        }
+      }
+    }
+  }
+
+  private void putLogicalTable(ZNRecord znRecord)
+      throws IOException {
+    LogicalTable logicalTable = LogicalTableUtils.fromZNRecord(znRecord);
+    String logicalTableName = logicalTable.getTableName();
+    _logicalTableNameMap.put(logicalTableName, logicalTable);
+  }
+
+  private void removeLogicalTable(String path) {
+    _propertyStore.unsubscribeDataChanges(path, _zkLogicalTableChangeListener);
+    String logicalTableName = path.substring(LOGICAL_TABLE_PATH_PREFIX.length());
+    _logicalTableNameMap.remove(logicalTableName);
+  }
+
+  private void notifyLogicalTableChangeListeners() {
+    if (!_logicalTableChangeListeners.isEmpty()) {
+      List<LogicalTable> logicalTables = new ArrayList<>(_logicalTableNameMap.values());
+      for (LogicalTableChangeListener listener : _logicalTableChangeListeners) {
+        listener.onChange(logicalTables);
+      }
+    }
+  }
+
+  @Nullable
+  public LogicalTable getLogicalTable(String logicalTableName) {
+    return _logicalTableNameMap.get(logicalTableName);
+  }
+
+  public boolean isLogicalTable(String logicalTableName) {
+    return _logicalTableNameMap.containsKey(logicalTableName);
+  }
+
+  public List<LogicalTable> getLogicalTables() {
+    return new ArrayList<>(_logicalTableNameMap.values());
+  }
+
+  public Schema getLogicalTableSchema(String logicalTableName) {
```

Comment:
We want to first lookup the schema using raw logicalTableName, then fallback to the first physical table.
@vrajat Do we want to enforce schema for logical table?

Reply:
> We want to first lookup the schema using raw logicalTableName, then fallback to the first physical table.

This would helpful when logical table has a schema enforced by the system. Otherwise we would not find schema same name as logical table. 

We can consider having an optional field to accept the schema from the user otherwise fallback to the first physical table schema.

@vrajat please check and comment.

Reply:
I prefer enforcing a schema. If we learn that it is overkill we can relax that requirement in a subsequent version. Adding a restriction is much harder. @Jackie-Jiang I'll first add this to the design doc ? We can discuss the pros/cons in the OSS slack channel. IIRC you had some concerns that RENAME will be harder.

Reply:
okay. I will not include the schema in this PR and we can add it after the discussion is concluded.